File: src/licitimart/ingestao/extracao.py
"""Extracao nativa de texto (RF-002) -- so PDF/DOCX nato-digital. Nenhuma
chamada de OCR paga aqui; PDF escaneado sem camada de texto vira
"requer_ocr" explicito (RF-020), nunca escondido nem inventado. Motor de
OCR (Document AI/Textract) fica para quando houver decisao/orcamento --
ver src/licitimart/ocr/README.md.
"""
import io
import logging

from docx import Document
from pypdf import PdfReader

logger = logging.getLogger(__name__)

# Media de caracteres por pagina abaixo disso = sinal classico de PDF
# escaneado sem camada de texto (uma pagina de edital real tem
# tipicamente centenas a milhares de caracteres). Numero conservador --
# prefere marcar "requer_ocr" a mais do que menos, RNF-012.
LIMIAR_CHARS_POR_PAGINA = 40


def extrair_texto_pdf(conteudo: bytes) -> tuple[str, str, int]:
    """Retorna (texto, status, paginas). status in ('extraido_nativo', 'requer_ocr', 'erro')."""
    try:
        leitor = PdfReader(io.BytesIO(conteudo))
        paginas = len(leitor.pages)
        textos = [p.extract_text() or "" for p in leitor.pages]
    except Exception as exc:
        logger.warning("erro extraindo PDF: %s", exc)
        return "", "erro", 0

    texto = "\n".join(textos)
    media_por_pagina = len(texto) / paginas if paginas else 0
    if media_por_pagina < LIMIAR_CHARS_POR_PAGINA:
        return texto, "requer_ocr", paginas
    return texto, "extraido_nativo", paginas


def extrair_texto_docx(conteudo: bytes) -> tuple[str, str, int]:
    """DOCX nao tem conceito de "pagina" no arquivo em si -- paginas
    retorna 0 (nao aplicavel). status "requer_ocr" nao existe para DOCX
    (nao ha DOCX escaneado); só "extraido_nativo" ou "erro"."""
    try:
        documento = Document(io.BytesIO(conteudo))
        texto = "\n".join(p.text for p in documento.paragraphs)
    except Exception as exc:
        logger.warning("erro extraindo DOCX: %s", exc)
        return "", "erro", 0
    return texto, "extraido_nativo", 0


def extrair_texto(conteudo: bytes, titulo: str) -> tuple[str, str, int]:
    """Despacha pelo conteudo real (magic bytes), nao pelo nome do
    arquivo -- achado real (Fase K, 10/09/2026): o campo "titulo" que o
    PNCP retorna nem sempre tem extensao (ex.: "EDITAL", sem ".pdf"),
    mesmo quando o arquivo E um PDF de verdade. Confiar só na extensao do
    nome gerava "erro" falso para documento perfeitamente extraivel."""
    if conteudo[:4] == b"%PDF":
        return extrair_texto_pdf(conteudo)
    if conteudo[:4] == b"PK\x03\x04":
        # docx e um zip -- tenta abrir como docx; se nao for (zip
        # generico, ex. anexo .zip de verdade), extrair_texto_docx
        # levanta e retorna "erro" honesto, nao finge sucesso.
        return extrair_texto_docx(conteudo)
    logger.info("formato nao suportado para extracao nativa (titulo=%r, primeiros bytes=%r)", titulo, conteudo[:8])
    return "", "erro", 0

File: src/licitimart/web/src/lib/pncp/extracao.ts
// Fase N: extração nativa de PDF/DOCX equivalente a
// src/licitimart/ingestao/extracao.py, portada para Node -- mesma regra:
// PDF escaneado sem camada de texto vira "requer_ocr" explícito, nunca
// inventado. Nenhuma chamada de OCR paga aqui.
import { PDFParse } from "pdf-parse";
import mammoth from "mammoth";

// Media de caracteres por pagina abaixo disso = sinal classico de PDF
// escaneado sem camada de texto -- mesmo limiar de extracao.py
// (LIMIAR_CHARS_POR_PAGINA), para os dois runtimes concordarem sobre o
// mesmo documento.
const LIMIAR_CHARS_POR_PAGINA = 40;

export interface ResultadoExtracao {
  texto: string;
  status: "extraido_nativo" | "requer_ocr" | "erro";
  paginas: number;
  offsets: number[];
}

async function extrairTextoPdf(conteudo: Buffer): Promise<ResultadoExtracao> {
  let parser: PDFParse | null = null;
  try {
    parser = new PDFParse({ data: conteudo });
    const resultado = await parser.getText();
    const textos = resultado.pages.map((p) => p.text ?? "");
    const paginas = resultado.total;

    // Mesmo calculo de offsets de extracao.py: offsets[i] e o indice (em
    // caracteres) onde a pagina i comeca dentro de `texto`.
    const offsets: number[] = [];
    let cursor = 0;
    for (const t of textos) {
      offsets.push(cursor);
      cursor += t.length + 1; // +1 pelo "\n" do join abaixo
    }
    const texto = textos.join("\n");

    const mediaPorPagina = paginas ? texto.length / paginas : 0;
    const status = mediaPorPagina < LIMIAR_CHARS_POR_PAGINA ? "requer_ocr" : "extraido_nativo";
    return { texto, status, paginas, offsets };
  } catch (exc) {
    console.warn("erro extraindo PDF:", exc);
    return { texto: "", status: "erro", paginas: 0, offsets: [] };
  } finally {
    await parser?.destroy();
  }
}

async function extrairTextoDocx(conteudo: Buffer): Promise<ResultadoExtracao> {
  // DOCX nao tem conceito de "pagina" no arquivo em si -- paginas e
  // offsets voltam vazios (nao aplicavel, nunca fingido), mesma
  // convencao do lado Python.
  try {
    const { value } = await mammoth.extractRawText({ buffer: conteudo });
    return { texto: value, status: "extraido_nativo", paginas: 0, offsets: [] };
  } catch {
    return { texto: "", status: "erro", paginas: 0, offsets: [] };
  }
}

// Despacha pelo conteudo real (magic bytes), nao pelo nome do arquivo --
// mesmo achado da Fase K: o "titulo" do PNCP nem sempre tem extensao.
export async function extrairTexto(conteudo: ArrayBuffer): Promise<ResultadoExtracao> {
  const buffer = Buffer.from(conteudo);
  if (buffer.length >= 4 && buffer.subarray(0, 4).toString("latin1") === "%PDF") {
    return extrairTextoPdf(buffer);
  }
  if (buffer.length >= 4 && buffer[0] === 0x50 && buffer[1] === 0x4b && buffer[2] === 0x03 && buffer[3] === 0x04) {
    return extrairTextoDocx(buffer);
  }
  return { texto: "", status: "erro", paginas: 0, offsets: [] };
}
